build/Shared/EqualityUtility.cs

```diff
@@ -14,48 +14,72 @@ namespace NuGet.Shared
+                if (sourceItems.TryGetValue(key, out int count))
+                {
+                    sourceItems[key] = count + 1;
+                }
+                else
+                {
+                    sourceItems[key] = 1;
+                }
+            }
+
+            int otherCount = 0;
+            foreach (TSource current in other)
+            {
+                ++otherCount;
+                TKey key = keySelector(current);
+                if (!sourceItems.TryGetValue(key, out int count) || count <= 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    sourceItems[key] = count - 1;
+                }
+            }
+
+            return selfCount == otherCount;
         }
 
         /// <summary>
-        /// Compares two collections for equality, ordered according to the specified key and optional comparer. Handles null values gracefully.
+        /// Compares two collections for equality using an optional comparer. Handles null values gracefully.
         /// </summary>
         /// <typeparam name="TSource">The type of the list</typeparam>
-        /// <typeparam name="TKey">The type of the sorting key</typeparam>
+        /// <typeparam name="TKey">The type used for equality checking</typeparam>
         /// <param name="self">This list</param>
         /// <param name="other">The other list</param>
         /// <param name="keySelector">The function to extract the key from each item in the list</param>
-        /// <param name="orderComparer">An optional comparer for comparing keys</param>
-        /// <param name="sequenceComparer">An optional comparer for sequences</param>
-        internal static bool OrderedEquals<TSource, TKey>(this ICollection<TSource>? self, ICollection<TSource>? other, Func<TSource, TKey> keySelector, IComparer<TKey>? orderComparer = null, IEqualityComparer<TSource>? sequenceComparer = null)
+        /// <param name="equalityComparer">An optional comparer for sequences</param>
+        internal static bool ElementsEqual<TSource, TKey>(this ICollection<TSource>? self, ICollection<TSource>? other, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? equalityComparer = null) where TKey : notnull
         {
-            Debug.Assert(orderComparer != null || typeof(TKey) != typeof(string), "Argument " + "orderComparer" + " must be provided if " + "TKey" + " is a string.");
-            Debug.Assert(sequenceComparer != null || typeof(TSource) != typeof(string), "Argument " + "sequenceComparer" + " must be provided if " + "TSource" + " is a string.");
+            Debug.Assert(equalityComparer != null || typeof(TKey) != typeof(string), "Argument " + nameof(equalityComparer) + " must be provided if " + nameof(TKey) + " is a string.");
 
-            bool identityEquals;
-            if (TryIdentityEquals(self, other, out identityEquals))
+            if (TryIdentityEquals(self, other, out bool identityEquals))
```

Comment:
this is just an opinion, not an ask to make chnages, but for PRs that are heavy on changes like this one, some of the non-functional changes make it more challenging to review.